### PR TITLE
[eas-cli] managed versions support

### PR DESCRIPTION
**Pull request by wkozyra95**
# Checklist

- [ ] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary. You can comment this pull request with `/changelog-entry [breaking-change|new-feature|bug-fix|chore] [message]` and CHANGELOG.md will be updated automatically.

# Why

Autoincrement versions when building with version source set to remote. This is the last part to make it work, but before this PR lands I still need to update docs and add new runtime version policy

# How

Android
20220725_11h44m15s_grim
iOS
20220725_11h44m59s_grim

# Test Plan

run few build for managed and bare projects

### CODE REVIEW

File: packages/eas-cli/src/build/android/version.ts [hunk 1]
```diff
@@ -165,3 +169,53 @@ async function writeBuildGradleAsync({
   const buildGradlePath = AndroidConfig.Paths.getAppBuildGradleFilePath(projectDir);
   await fs.writeFile(buildGradlePath, buildGradle);
 }
+
+export async function updateToNextVersionCodeAsync({
```
Comment (dsokal) [inline packages/eas-cli/src/build/android/version.ts:173]:
```suggestion
export async function bumpNextVersionCodeAsync({
```

File: packages/eas-cli/src/build/android/build.ts [hunk 2]
```diff
@@ -60,8 +62,18 @@ This means that it will most likely produce an AAB and you will not be able to i
   }
 
   const applicationId = await getApplicationIdAsync(ctx.projectDir, ctx.exp, gradleContext);
+  const overrideVersionCode =
```
Comment (dsokal) [inline packages/eas-cli/src/build/android/build.ts:65]:
```suggestion
  const versionCodeOverride =
```

File: packages/eas-cli/src/build/android/version.ts [hunk 3]
```diff
@@ -165,3 +169,53 @@ async function writeBuildGradleAsync({
...
+    projectId,
+    AppPlatform.Android,
+    applicationId
+  );
+
+  const localVersions = await maybeResolveVersionsAsync(projectDir, exp, buildProfile);
+  let currentBuildVersion: string;
+  if (!remoteVersions?.buildVersion && !localVersions.appBuildVersion) {
+    Log.error(
+      `Remote versions are not configured and we were not able to read the current version from the local project. Use "eas build:version:set" to initialize remote versions.`
```
Comment (dsokal) [inline packages/eas-cli/src/build/android/version.ts:196]: we -> EAS CLI

Comment (dsokal) [reply packages/eas-cli/src/build/android/version.ts:196]: local project -> your project (?)

File: packages/eas-cli/src/build/android/version.ts [hunk 4]
```diff
@@ -165,3 +169,53 @@ async function writeBuildGradleAsync({
...
+  applicationId: string;
+  buildProfile: BuildProfile<Platform.ANDROID>;
+}): Promise<string> {
+  const remoteVersions = await AppVersionQuery.latestVersionAsync(
+    projectId,
+    AppPlatform.Android,
+    applicationId
+  );
+
+  const localVersions = await maybeResolveVersionsAsync(projectDir, exp, buildProfile);
```
Comment (dsokal) [inline packages/eas-cli/src/build/android/version.ts:192]: Maybe move this function outside the `metadata.ts` module.

File: packages/eas-cli/src/build/android/version.ts [hunk 5]
```diff
@@ -165,3 +169,53 @@ async function writeBuildGradleAsync({
...
+      `Remote versions are not configured and we were not able to read the current version from the local project. Use "eas build:version:set" to initialize remote versions.`
+    );
+    throw new Error('Remote versions are not configured.');
+  } else if (!remoteVersions?.buildVersion && localVersions.appBuildVersion) {
+    Log.warn(
+      'No remote versions are configured for this project, versionCode will be initialized based on the value from the local project.'
+    );
+    currentBuildVersion = localVersions.appBuildVersion;
+  } else {
+    assert(remoteVersions?.buildVersion);
```
Comment (dsokal) [inline packages/eas-cli/src/build/android/version.ts:205]: I think you can get rid of this assert by doing sth like:
```
   if (!remoteVersions?.buildVersion) {
    if () {
    } else {
    }
   } else {
   }
```

File: packages/eas-cli/src/build/context.ts [hunk 6]
```diff
@@ -16,13 +16,15 @@ export type CommonContext<T extends Platform> = Omit<BuildContext<T>, 'android'
 export interface AndroidBuildContext {
   applicationId: string;
   gradleContext?: GradleBuildContext;
+  overrideVersionCode?: string;
```
Comment (dsokal) [inline packages/eas-cli/src/build/context.ts:19]:
```suggestion
  versionCodeOverride?: string;
```

File: packages/eas-cli/src/build/context.ts [hunk 7]
```diff
@@ -16,13 +16,15 @@ export type CommonContext<T extends Platform> = Omit<BuildContext<T>, 'android'
...
 }
 
 export interface IosBuildContext {
   bundleIdentifier: string;
   applicationTarget: Target;
   targets: Target[];
   xcodeBuildContext: XcodeBuildContext;
+  overrideBuildNumber?: string;
```
Comment (dsokal) [inline packages/eas-cli/src/build/context.ts:27]:
```suggestion
  buildNumberOverride?: string;
```

File: packages/eas-cli/src/commands/build/version/set.ts [hunk 8]
```diff
@@ -49,13 +48,13 @@ export default class BuildVersionSetView extends EasCommand {
 
     const platform = await selectPlatformAsync(flags.platform);
     const easJsonReader = new EasJsonReader(projectDir);
-    await ensureRemoteVersionPolicyAsync(projectDir, easJsonReader);
+    await ensureRemoteVersionSourceAsync(projectDir, easJsonReader);
```
Comment (dsokal) [inline packages/eas-cli/src/commands/build/version/set.ts:51]: ensure ... what ...?

File: packages/eas-cli/src/build/ios/version.ts [hunk 9]
```diff
@@ -245,3 +249,60 @@ export function evaluateTemplateString(
     }
   });
 }
+
+export async function updateToNextBuildNumberAsync({
```
Comment (dsokal) [inline packages/eas-cli/src/build/ios/version.ts:253]:
```suggestion
export async function bumpBuildNumberAsync({
```

File: packages/eas-cli/src/build/ios/version.ts [hunk 10]
```diff
@@ -245,3 +249,60 @@ export function evaluateTemplateString(
...
+  );
+  const localShortVersion = await readShortVersionAsync(
+    projectDir,
+    exp,
+    applicationTarget.buildSettings ?? {}
+  );
+  let currentBuildVersion: string;
+  if (!remoteVersions?.buildVersion && !localBuildNumber) {
+    Log.error(
+      `Remote versions are not configured and we were not able to read the current version from the local project. Use "eas build:version:set" to initialize remote versions.`
```
Comment (dsokal) [inline packages/eas-cli/src/build/ios/version.ts:283]: we -> EAS CLI

Comment (dsokal) [reply packages/eas-cli/src/build/ios/version.ts:283]: local project -> your project.

File: packages/eas-cli/src/build/ios/version.ts [hunk 11]
```diff
@@ -245,3 +249,60 @@ export function evaluateTemplateString(
...
+    );
+    throw new Error('Remote versions are not configured.');
+  } else if (!remoteVersions?.buildVersion && localBuildNumber) {
+    Log.warn(
+      'No remote versions are configured for this project, buildNumber will be initialized based on the value from the local project.'
+    );
+    currentBuildVersion = localBuildNumber;
+  } else {
+    assert(remoteVersions?.buildVersion);
```
Comment (dsokal) [inline packages/eas-cli/src/build/ios/version.ts:292]: Same comment as for android.

File: packages/eas-cli/src/commands/build/version/set.ts (hunk 8 above)
Comment (wkozyra95) [reply packages/eas-cli/src/commands/build/version/set.ts:51]: `ensureVersionSourceIsRemoteAsync`?

File: packages/eas-cli/src/build/android/version.ts (hunk 4 above)
Comment (wkozyra95) [reply packages/eas-cli/src/build/android/version.ts:192]: not sure what you mean, this function is not defined in metadata.ts but there is a function there with the same name

**Review by dsokal — APPROVED** [review]

File: packages/eas-cli/src/build/android/version.ts [hunk 12]
```diff
@@ -170,7 +169,11 @@ async function writeBuildGradleAsync({
   await fs.writeFile(buildGradlePath, buildGradle);
 }
 
-export async function updateToNextVersionCodeAsync({
+/**
+ * Returns buildNumber that will be used for the next build. If current build profile
+ * has an 'autoIncrement' option set, it increments the version on server.
```
[the remark this answers is not in the document]
Comment (wkozyra95) [reply packages/eas-cli/src/build/android/version.ts:174]: yes there might be gaps in build numbers, this is expected

File: packages/eas-cli/src/build/android/version.ts [hunk 13]
```diff
@@ -165,3 +168,61 @@ async function writeBuildGradleAsync({
...
+      );
+      throw new Error('Remote versions are not configured.');
+    }
+  }
+  if (!buildProfile.autoIncrement) {
+    return currentBuildVersion;
+  }
+
+  const nextBuildVersion = getNextVersionCode(currentBuildVersion);
+  Log.log(`Incrementing versionCode ${currentBuildVersion} -> ${nextBuildVersion}.`);
```
Comment (brentvatne) [inline packages/eas-cli/src/build/android/version.ts:217]: let's use ora here so we show a spinner when it's happening and a checkmark when it's done. text should read "Incrementing" when the request is in flight and "Incremented" when completed

**Review by brentvatne — APPROVED** [review]

**Comment by brentvatne** [comment]
i tried running this locally on a new blank project w/ the following eas.json:

```json
{
  "cli": {
    "version": ">= 0.56.0",
    "appVersionSource": "remote"
  },
  "build": {
    "preview": {
      "distribution": "internal",
      "ios": {
        "autoIncrement": true,
        "simulator": true
      }
    }
  }
}
```

i ran `easd build -p ios --profile preview`

```
╰─$ easd build -p ios --profile preview                                                                                                                        130 ↵
✔ Linked to project @brents/example

📝  iOS Bundle Identifier Learn more
✔ What would you like your iOS bundle identifier to be? … com.brents.example
No remote versions are configured for this project, buildNumber will be initialized based on the value from the local project.
✔ Incremented buildNumber 1 -> 2.

Bumping expo.ios.buildNumber from 1 to 2

Compressing project files and uploading to EAS Build. Learn more
✔ Uploaded to EAS 3s

....
```

It added `expo.ios.buildNumber` to **app.json** even though it didn't exist yet, and incremented it. I believe what we wanted here was to skip incrementing it locally when the source is set to remote.

**Comment by wkozyra95** [comment]
> It added expo.ios.buildNumber to app.json even though it didn't exist yet, and incremented it. I believe what we wanted here was to skip incrementing it locally when the source is set to remote.

yes, that was the intended behavior, It worked like that on screenshots that are in the description, not sure how I introduced that regression, maybe bad rebase.

**Review by dsokal — APPROVED** [review]

**Pull request merged**